### PR TITLE
Y10k fix

### LINKED ISSUE #2140: Y10K bug:  Restart file names mangled after model crosses year 10,000 boundary
#
  I am running aquaplanet.  After crossing the Y10K model time boundary some filenames
 are mangled.  The cam.h0 and cam.r file names were *correctly* configured, for example:

  cam4_aqua_4x5.cam.h0.10031-01-01-00000.nc
  cam4_aqua_4x5.cam.r.10031-01-01-00000.nc
  cam4_aqua_4x5.cam.rs.10031-01-01-00000.nc

But the following filenames were mangled because of an apparent 4 character restriction:
```
cam4_aqua_4x5.cpl.hi.****-01-01-01800.nc
cam4_aqua_4x5.cpl.r.****-01-01-00000.nc
cam4_aqua_4x5.docn.rs1.****-01-01-00000.bin
```
  The model is still successfully able to restart, however, using these mangled files

                 Jerry Olson

### PR DESCRIPTION
**Pull request by jedwards4b**
Allow model years > 9999  

Test suite: scripts_regression_tests.py with new test mod
Test baseline: 
Test namelist changes: 
Test status: bit for bit

Fixes #2140 

User interface changes?: 

Update gh-pages html (Y/N)?:

Code review:

### CODE REVIEW

**Comment by billsacks** [comment]
@jedwards4b thanks for addressing this. I'll look more closely later, but a quick comment for now: I see you have allowed an extra 0, allowing us to go to year 99,999. For ice sheet and paleo runs, it is not uncommon to go beyond 100,000. Would it work to allow still another digit - or even 2 or 3?

**Comment by jedwards4b** [comment]
Actually six digits are allowed:  999,999
The only 5 digit limit is in STOP_YMD and that can be overwritten with an xml change.

**Comment by jgfouca** [comment]
Wow, didn't realize anyone was doing simulations out that far!

**Comment by mvertens** [comment]
Its needed for paleo-climate runs.

**Comment by billsacks** [comment]
It looks like some changes are needed in the scripts. Here are some assumptions that restart and history file names use 4-digit years:

```
$ git grep -n '\[0-9\]{4}'
scripts/lib/CIME/case_st_archive.py:74:    re_formats = [r"[0-9]{4}-[0-9]{1,2}-[0-9]{1,2}_[0-9]{1,2}\.[0-9]{1,2}\.[0-9]{1,2}", # yyyy-mm-dd_hh.MM.ss
scripts/lib/CIME/case_st_archive.py:75:                  r"[0-9]{4}-[0-9]{1,2}-[0-9]{1,2}[\-_][0-9]{1,5}",                     # yyyy-mm-dd_sssss
scripts/lib/CIME/case_st_archive.py:76:                  r"[0-9]{4}-[0-9]{1,2}-[0-9]{1,2}",                                    # yyyy-mm-dd
scripts/lib/CIME/case_st_archive.py:77:                  r"[0-9]{4}[\-\.][0-9]{1,2}",                                          # yyyy-mm
scripts/lib/CIME/hist_utils.py:131:            m = re.search("(.+)_[0-9]{4}(.+.nc)",normalized_name)
scripts/lib/CIME/hist_utils.py:265:    multiinst_regex = re.compile(r'.*%s[^_]*(_[0-9]{4})[.]h.?[.][^.]+?[.]nc' % model)
scripts/lib/CIME/hist_utils.py:356:    ext_regex = re.compile(r'.*%s[^_]*_?([0-9]{4})?[.](h.?)([.].*[^.])?[.]nc' % model)
scripts/lib/CIME/hist_utils.py:409:            m = re.search("(.*%s.*)_([0-9]{4})(.h.*)" % model, baseline)
```

I'm not sure if that grep would have picked everything up.

It would be good to run a test that covers short-term archiving to see if something else is missing.

**Comment by billsacks** [comment]
Changes are also needed to this function in seq_io_mod: `seq_io_date2yyyymmdd`.

Actually, it looks like that function is going to be problematic, and there are probably similar problems elsewhere: We apparently represent dates with 4-byte integers in some places. This allows for a max of 2147483647. If this is done like I think it is, then I think this would allow years up to 214748.

If that's right, then my suggestion of allowing 6-digit years won't work everywhere, and maybe we should go back to 5-digit years for now. If so, you could revert your recent commit, and disregard my comments (coming soon in my review) pointing out other places things need to be changed to allow 6-digit years.

File: src/drivers/mct/shr/seq_timemgr_mod.F90 [hunk 1]
```diff
@@ -528,7 +528,7 @@ subroutine seq_timemgr_clockInit(SyncClock, nmlfile, restart, restart_file, pioi
        endif
 
        if ( stop_ymd < 0) then
-          stop_ymd = 99990101
+          stop_ymd = 999990101
```
[the remark this answers is not in the document]
Comment (billsacks) [reply src/drivers/mct/shr/seq_timemgr_mod.F90]: Never mind, that won't work because of limits on int sizes.

**Comment by jedwards4b** [comment]
Updating the fortran to accept years up to 999999 was not too difficult and I've updated the PR with that change.   A problem remains however in the python - python datetime only supports up to 9999 - and I haven't found any easy way to extend it.   It seems like a pretty arbitrary value to me.
I would like to suggest accepting this PR and opening another issue on python support - which appears to be limited to file case_st_archive.py

**Comment by jedwards4b** [comment]
Opened #2162 against python datetime issue.

**Comment by jedwards4b** [comment]
@billsacks your comments on hist_utils.py are incorrect, that is a match to model instance.

**Comment by billsacks** [comment]
Okay, with the changes I made, I'm happy with this. @jedwards4b thanks a lot for taking this on - this will be a big help for paleo runs. @jgfouca I feel this is ready whenever you want to merge it.

**Pull request merged**